worldeditadditions_commands/commands/selectors/smake.lua

```diff
@@ -0,0 +1,141 @@
+		-- Check base if base is present and if so valid.
+		if base then
+			if base:match("[xyz]") then -- ensure correct base syntax
+				base = base:match("[xyz]")
+			else
+				return false, "Error: Invalid base \""..base.."\". Expected \"x\", \"y\" or \"z\"."
+			end
+		end
+		
+		-- Resolve target then mode (in that order incase mode is target).
+		if not targ then -- If no target set to default (xz)
+			targ = "xz"
+		elseif targ:match("[xyz]+") then -- ensure correct target syntax
+			targ = table.concat(wea.tochars(targ:match("[xyz]+"),true,true))
+		else
+			return false, "Error: Invalid <target> \""..targ.."\". Expected \"x\" and or \"y\" and or \"z\"."
+		end
+		
+		if mode == "average" then -- If mode is average set to avg
+			mode = "avg"
+		elseif mode:match("[xyz]+") then -- If target is actually base set vars to correct values.
+			base, targ, mode = targ:sub(1,1), table.concat(wea.tochars(mode:match("[xyz]+"),true,true)), false
+		elseif not modeSet[mode] and not base then -- If mode is invalid and base isn't present throw error
+			return false, "Error: Invalid <mode> \""..mode.."\". Expected \"grow\", \"shrink\", or \"average\"/\"avg\"."
+		end
+		
+		if base then
+			if oper ~= "equal" then base = false -- If operation isn't equalize we don't need <base>
+			elseif targ:match(base) then -- Else check that base is not in target and return error if it is
+				return false, "Error: <base> ("..base..") cannot be included in <target> ("..targ..")."
+			end
+		end
+		
+		-- Check if operator is valid
+		if not operSet[oper] then
+			return false, "Error: Invalid operator \""..oper.."\". Expected \"odd\", \"even\" or \"equal\"."
+		end
+		
+		if false then
+			return false, "<operator>: " .. oper .. ", <mode>: " .. tostring(mode) .. ", <target>: " .. tostring(targ) .. ", <base>: " .. tostring(base)
+		end
+		return true, oper, mode, targ, base
+	end,
+	func = function(name, oper, mode, targ, base)
+		local p1, p2, eval = vector.new(worldedit.pos1[name]), vector.new(worldedit.pos2[name]), function(int) return int or 0 end
+		local delta, _tl, targ, _m = vector.subtract(p2,p1), #targ, wea.tocharset(targ), 0 -- local delta equation: Vd(a) = V2(a) - V1(a)
+		
+		-- set _m to the max, min or mean of the target axes depending on mode (_tl is the length of targ) or base if it exists
+		if base then _m = delta[base]
+		elseif mode == "avg" then
+			for k,v in pairs(targ) do _m = _m + math.abs(delta[k]) end
+			_m = _m / _tl
+		elseif mode == "grow" then
+			for k,v in pairs(targ) do if math.abs(delta[k]) > _m then _m = math.abs(delta[k]) end end
+		else
+			-- Take output of next(targ), put it in a table, wrap the table in brackets to force evlauation so that
+			-- we can call the first element of that table to serve as the key for a call to delta.
+			_m = delta[({next(targ)})[1]]
+			for k,v in pairs(targ) do if math.abs(delta[k]) < _m then _m = math.abs(delta[k]) end end
+		end
+		
+		if oper == "even" then
+			eval = function(int)
```

Comment:
I'm cautious about nested functions like this. Could they be moved to the top of the file please (e.g. `local function NAME_HERE`)?

Same for the other nested functions below.

Reply:
Sorry but no. The fact that //smake is even a little bit efficient in it's execution is because of this design choice. Instead of declaring 3 similar local functions at the top and then having a bit if tree in the evaluation loop, I declare one local function at the top and then dynamically modify it with the if tree and then call the modified function from the evaluation loop.

I know it's not an obvious implementation (it took me a week and a half to come up with) but it shaves down the load at runtime.

Reply:
You could still do this:

```lua
local function a()
    -- ....
end
local function b()
    -- ....
end

-- ....
local eval = a
if foo then eval = b end
```

...unless I'm misunderstanding something?

We can always keep it as-is though.

Reply:
> ...unless I'm misunderstanding something?

I think you have a pretty good handle on what's going on except for that you're still declaring 3 functions instead of just declaring one and editing it once. 2 won't be used but will still take resources (not very much but still...).

The big thing for me is that by having that big, almost repetitive chunk of code in the if tree I can prevent 2/3rds of the tree from ever evaluating when the function  is called and that's more efficient.

Reply:
I see. I'm not convinced it's a noticeable performance difference given that this is only executed once per command invocation (i.e. unlike in a triple-nested for loop that's scanning a large area of nodes, but let's let it be for now then.